feat: Support Helm value files from OCI repositories (#24376)

faxioman opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

## Summary

This PR adds support for using OCI repositories as reference sources for Helm value files in multi-source applications. This feature was requested in issue #24376 to enable organizations using OCI registries as their primary artifact store to manage Helm values alongside their container images and charts.

## Why is this PR necessary?

Currently, Argo CD only supports Git repositories as reference sources for Helm value files. Many organizations are adopting OCI registries for all their artifacts, including Helm charts and configuration files. This limitation forces them to maintain separate Git repositories solely for value files, adding unnecessary complexity to their GitOps workflows.

## Changes

### Core Functionality
- Extended `resolveReferencedSources` to handle both Git and OCI repositories
- Added OCI client getter and resolution logic for OCI-based value files  
- Modified manifest generation to support OCI paths alongside Git paths
- Updated Helm template functions to accept OCI repository references

### Key Implementation Details
- **URL Handling**: OCI repository URLs are preserved as-is (no normalization applied, unlike Git URLs)
- **Chart Field Support**: Chart fields are now allowed for OCI ref sources (previously only supported for Git)
- **Path Resolution**: Added `getResolvedOCIRefValueFile` to properly extract and resolve value file paths from OCI artifacts
- **Context Propagation**: Added context parameter to functions that interact with OCI registries

### Documentation
- Updated [user guide](docs/user-guide/multiple_sources.md) to document OCI repository support for Helm values
- Added example configurations showing OCI ref usage

### Testing
- Added unit tests for OCI value file resolution ([`TestHelmChartReferencingOCIValues`](reposerver/repository/repository_test.go))
- Added error handling tests for invalid OCI references
- Added E2E test [`TestMultiSourceAppWithOCIRefValues`](test/e2e/oci_test.go)
- Created test fixtures with sample OCI value files

## Usage Example

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
spec:
  sources:
  - repoURL: https://github.com/my-org/my-app
    targetRevision: main
    path: manifests
    
  - repoURL: oci://registry.example.com/charts
    chart: my-chart
    targetRevision: 1.0.0
    helm:
      valueFiles:
      - $ref/values.yaml  # Reference to OCI values
      
  - repoURL: oci://registry.example.com/config/app-values
    targetRevision: 1.0.0
    ref: ref  # Reference name for OCI repository
```

## Benefits
- **Unified Configuration Management**: Store Helm values in OCI registries alongside container images and charts
- **Version Control**: Leverage OCI's native versioning for configuration management
- **Consistency**: Use the same `$ref` syntax for both Git and OCI sources
- **Flexibility**: Support for environments where OCI registries are the primary artifact store

## Backward Compatibility
This change is fully backward compatible. Existing applications using Git repositories for value files continue to work without any modifications.

## Note on Implementation Status
- Backend support is fully implemented with this PR
- CLI/UI updates may be submitted in a follow-up PR if additional exposure is needed

Fixes #24376

faxioman committed:
Merge branch 'master' into master-pr

faxioman committed:
feat: Support Helm value files from OCI repositories

This commit adds support for using OCI repositories as reference sources
for Helm value files in multi-source applications.

Key changes:
- Extended resolveReferencedSources to handle both Git and OCI repositories
- Added OCI client getter and resolution logic for OCI-based value files
- Modified manifest generation to support OCI paths alongside Git paths
- Updated Helm template functions to accept OCI repository references
- URL handling preserves OCI URLs without normalization (unlike Git URLs)
- Chart fields are now allowed for OCI ref sources
- Added getResolvedOCIRefValueFile to resolve value file paths from OCI artifacts
- Context parameter added to functions that interact with OCI registries
- Updated documentation to explain OCI repository support
- Added comprehensive tests for OCI value file resolution
- Fixed linter errors (gofumpt, perfsprint, revive, staticcheck)

Fixes #24376

Signed-off-by: Fabio Masini <fabio@rapidosoft.it>

faxioman committed:
Merge branch 'master' into master

reggie-k commented on reposerver/repository/repository.go:
```diff
@@ -564,16 +566,40 @@ func resolveReferencedSources(hasMultipleSources bool, source *v1alpha1.Applicat
+			normalizedRepoURL = refSourceMapping.Repo.Repo
+		} else {
+			// For Git repositories, normalize the URL
+			normalizedRepoURL = git.NormalizeGitURL(refSourceMapping.Repo.Repo)
```
Can we extract the repo URL normalization to a separate function? It will make the code more readable as this logic is used multiple times in the code and will help with unit testing.

ppapapetrou76 reviewed (commented):
The code looks hard to read and understand. 
I highly recommend to spend time and simplify it by introducing maybe some structs or extract code to unexported functions etc. 

Thanks

faxioman committed:
refactor: address review feedback on repository handling

- Add IsOCI() and NormalizeRepoURL() methods to Repository struct
- Extract value file resolution logic into dedicated ValueFileResolver
- Remove unused context parameter from helmTemplate
- Eliminate duplicate OCI URL checks using IsOCIURL() function

Signed-off-by: Fabio Masini <fabio@rapidosoft.it>

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

chansuke commented on reposerver/repository/repository.go:
```diff
@@ -1428,6 +1546,38 @@ func getResolvedRefValueFile(
+	}
+
+	// Remove first segment (the ref variable name) and resolve the path
+	pathStrings[0] = "" // Remove first segment. It will be inserted by pathutil.ResolveValueFilePathOrUrl.
```
should we guard `pathStrings[0] = ""` with a length check? 
an empty rawValueFile (or a leading slash) would make pathStrings would leave pathStrings empty and panic when we touch index 0?

chansuke reviewed (commented):
Added some comments

blakepettersson commented on reposerver/repository/repository.go:
```diff
@@ -564,16 +566,38 @@ func resolveReferencedSources(hasMultipleSources bool, source *v1alpha1.Applicat
+			normalizedRepoURL = git.NormalizeGitURL(refSourceMapping.Repo.Repo)
 		}
-		normalizedRepoURL := git.NormalizeGitURL(refSourceMapping.Repo.Repo)
+
```
I think we can just use 
```suggestion
		normalizedRepoURL := git.NormalizeGitURL(refSourceMapping.Repo.Repo)
```

This should be fine to normalize even with OCI URLs (all the logic in there is specific to Git itself and shouldn't affect OCI)

faxioman committed:
Add length check for pathStrings slice

Signed-off-by: Fabio Masini <fabio@rapidosoft.it>

faxioman committed:
Simplify URL normalization for ref sources

Signed-off-by: Fabio Masini <fabio@rapidosoft.it>

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

blakepettersson commented on reposerver/repository/repository.go:
```diff
@@ -810,71 +827,126 @@ func (s *Service) runManifestGenAsync(ctx context.Context, repoRoot, commitSHA,
-					if refSourceMapping.Chart != "" {
-						ch.errCh <- errors.New("source has a 'chart' field defined, but Helm charts are not yet not supported for 'ref' sources")
+					// Allow OCI ref sources with chart field, but still restrict Git ref sources with chart field
+					if refSourceMapping.Chart != "" && !v1alpha1.IsOCIURL(refSourceMapping.Repo.Repo) {
+						ch.errCh <- errors.New("source has a 'chart' field defined, but Helm charts are not yet supported for Git 'ref' sources")
 						return
 					}
-					normalizedRepoURL := git.NormalizeGitURL(refSourceMapping.Repo.Repo)
+
+					normalizedRepoURL := refSourceMapping.Repo.NormalizeRepoURL()
```
I don't think an OCI source with a chart field works in practice, have you tested this part?

faxioman replied:
You’re right.
My current use case uses an OCI source with path: ., so it doesn’t exercise the chart case.
Since the implementation also allows OCI ref sources with chart, I’ll add coverage for that syntax and verify that the OCI chart is actually materialized so $values/... can resolve files from it.

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

faxioman committed:
Merge branch 'master' into master

smartaquarius10 commented:
Hello @faxioman, hope u are doing great. 
Its a really cool enhancement. Any plans of releasing within this month. Actually, I am stuck with some scenario where in this feature would be really helpful. Thanks

blakepettersson commented on reposerver/repository/repository.go:
```diff
@@ -1263,7 +1335,11 @@ func helmTemplate(appPath string, repoRoot string, env *v1alpha1.Env, q *apiclie
 				// If the $-prefixed path appears to reference another source, do env substitution _after_ resolving the source
-				resolvedPath, err = getResolvedRefValueFile(p.Path, env, q.GetValuesFileSchemes(), referencedSource.Repo.Repo, gitRepoPaths)
+				if referencedSource.Repo.IsOCI() {
+					resolvedPath, err = getResolvedOCIRefValueFile(p.Path, env, q.GetValuesFileSchemes(), referencedSource.Repo.Repo, ociPaths)
+				} else {
+					resolvedPath, err = getResolvedRefValueFile(p.Path, env, q.GetValuesFileSchemes(), referencedSource.Repo.Repo, gitRepoPaths)
+				}
```
Rather than branching between OCI/Git for each of these invocations, can we please instead have one `getResolvedRefValueFile` that rules them all? Something like 

```golang

func getResolvedRefValueFile(
	rawValueFile string,
	env *v1alpha1.Env,
	allowedValueFilesSchemas []string,
	refSource *v1alpha1.RefTarget,
	gitRepoPaths utilio.TempPaths,
	ociPaths utilio.TempPaths,
) (pathutil.ResolvedFilePath, error) {
// internal branching goes here
}
```

And do the same for any callsites that differentiates between Git/OCI with values files? I think having the `value_file_resolver.go` is a good start. We can move anything values file related in there, since as it stands now it's hard to follow the flow since we need to jump between `value_file_resolver.go` and `repository.go`.

faxioman replied:
Makes sense. I’ll consolidate the Git/OCI branching inside getResolvedRefValueFile and pass both gitRepoPaths and ociPaths into it, so callers don’t need to know which backing store is used.

I’ll also move the remaining values-file-specific logic into value_file_resolver.go to keep repository.go easier to follow.

blakepettersson reviewed (changes requested):
I like the feature, but we need to consolidate this so it's easier to maintain.

wpjunior commented:
Hi @faxioman and @blakepettersson I made some adjusts that was requested here, also I did the merge from master, this is on my github fork: https://github.com/wpjunior/argo-cd/pull/1

blakepettersson commented:
Hi @wpjunior,

Can you make a PR to the argocd repo so we can take a closer look (and ensure that it passes tests, compiles etc)?

wpjunior commented:
> Hi @wpjunior,
> 
> Can you make a PR to the argocd repo so we can take a closer look (and ensure that it passes tests, compiles etc)?

Hi @blakepettersson, as soon as possible I will open it, first I am fixing the e2e tests on my fork, after that I will open it =)
